### PR TITLE
Optimize angle-based modis interpolation for dask

### PR DESCRIPTION
**Pull request by djhoese**
This was mostly discussed on slack, but the angle-based interpolation solution for MODIS interpolation is pretty slow and produces a lot of dask tasks. Additionally, it produces very odd chunking of the results for some reason (perhaps due to slicing and appending). This PR is an attempt at cleaning some of that up and trying to get some consistency between the angled-based and the "simple" interpolation.

 - [ ] Closes #xxxx 
 - [ ] Tests added 
 - [ ] Tests passed 
 - [ ] Passes ``git diff origin/main **/*py | flake8 --diff`` 
 - [ ] Fully documented

### CODE REVIEW

**Comment by djhoese** [comment]
Ok so this overall code performs better (twice as fast-ish), but the code is pretty gross. I couldn't use the class methods because everything was being passed to map_blocks so at least the way it was it wouldn't have worked. But now all the values that were class properties are basically passed to every method. I *could* create a class inside the map_blocks function :thinking: 

Thoughts @mraspaud?

**Comment by djhoese** [comment]
I also don't like the CPU usage:

image

This says to me that the processing is wasting time doing something else (memory allocation maybe?).

File: geotiepoints/modisinterpolator.py [hunk 1]
```diff
@@ -78,146 +79,239 @@ def compute_expansion_alignment(satz_a, satz_b, satz_c, satz_d):
...
+    fine_pixels_per_coarse_pixel,
+    arr,
+    fine_scan_length,
+):
+    arr = np.repeat(arr, fine_scan_length, axis=1)
+    arr = np.concatenate(
+        (arr[:, : fine_scan_length // 2, :], arr, arr[:, -(fine_scan_length // 2):, :]), axis=1
+    )
+    arr = np.repeat(arr.reshape((-1, coarse_scan_width - 1)), fine_pixels_per_coarse_pixel, axis=1)
+    return np.hstack((arr, arr[:, -fine_pixels_per_coarse_pixel:]))
```
[the remark this answers is not in the document]
Comment (mraspaud) [reply geotiepoints/modisinterpolator.py]: ok, this is definitely unreadable :)

File: geotiepoints/modisinterpolator.py [hunk 2]
```diff
@@ -78,83 +78,138 @@ def compute_expansion_alignment(satz_a, satz_b, satz_c, satz_d):
...
+                                 coarse_scan_width=self._coarse_scan_width)
+    @scanline_mapblocks
+    def _interpolate(
+        self,
+        lon1,
+        lat1,
+        satz1,
+        coarse_resolution=None,
+        fine_resolution=None,
+        coarse_scan_width=None,
```
Comment (mraspaud) [inline geotiepoints/modisinterpolator.py:137]: why do these need to be passed vs retrieve from `self`?

Comment (djhoese) [reply geotiepoints/modisinterpolator.py:137]: Read the note in the docstring. The decorator uses them to determine what the output shape and chunk size is going to be.

Comment (mraspaud) [reply geotiepoints/modisinterpolator.py:137]: that's a bit ugly :)

Comment (djhoese) [reply geotiepoints/modisinterpolator.py:137]: I could move the `__init__` logic into this method but then the `__init__` has no purpose. I could change this back into a function (again) and create the class inside the map_blocks'd function. Then this all makes sense.

**Comment by djhoese** [comment]
Regarding the small coverage loss, this seems to be mostly for error cases not being checked in the scan-aligned block decorator. I'd like to wait to add tests for these as I do more restructuring of these modules in #38.

File: geotiepoints/modisinterpolator.py [hunk 3]
```diff
@@ -84,177 +85,214 @@ def compute_expansion_alignment(satz_a, satz_b, satz_c, satz_d, scan_width):
...
-            x[-6] = 6
-            x[-5] = 7
-            x[-4] = 8
-            x[-3] = 9
-            x[-2] = 10
-            x[-1] = 11
-        else:
-            raise NotImplementedError("Can't interpolate if 5km tiepoints have less than 270 columns.")
-        return x, y
+class _Interpolator:
```
Comment (mraspaud) [inline geotiepoints/modisinterpolator.py:116]: Even if this is private, I think it should have Modis in its name since it is specific to that instrument.

**Review by mraspaud — COMMENTED** [review]
Looks good, just one comment.

File: geotiepoints/modisinterpolator.py (hunk 3 above)
Comment (djhoese) [reply geotiepoints/modisinterpolator.py:116]: I guess...but it is in the `modisinterpolator.py` module. If I change that would you be OK with a merge without another review?

Comment (mraspaud) [reply geotiepoints/modisinterpolator.py:116]: yes!

**Pull request merged**